packages/vaex-hdf5/vaex/hdf5/export.py

```diff
@@ -156,13 +158,17 @@ def export_hdf5(dataset, path, column_names=None, byteorder="=", shuffle=False,
                 #     data_shape = column.bytes.shape
                 #     indices_shape = column.indices.shape
                 # else:
-
-                byte_length = dataset[column_name].str.byte_length().sum(selection=selection)
+                
+                byte_length = dataset[column_name].str.byte_length()
+                gc.collect()
+                byte_length = byte_length.sum(selection=selection)
+                gc.collect()
```

Comment:
@nf78 thanks for looking into this! Do you have an idea why Python does not release the memory automatically in this case, and therefore we have to hardcode the call to `gc.collect()`? Garbage collection has performance implications, and is supposed to be maintained by Python itself. So if we have to add this to the library code, does it mean that we should also recommend Vaex users to call `gc.collect()` after every (or some) aggregations?

What is the root cause of this problem? Can we make sure that this aggregation allocates smaller chunks of memory, that would be garbage-collected automatically?

Reply:
Thanks @byaminov I'm really happy to help !

 Before I answer your questions, I would like to make a little introduction here:

**Memory management**

Unlike many other languages, Python does not necessarily release the memory back to the Operating System. Instead, it has a dedicated object allocator for small objects, which keeps some already allocated memory for further use in the future. The amount of memory that Python holds depends on the usage patterns. In some cases, all allocated memory could be released only when Python process terminates.

**Garbage collection algorithms**
Standard CPython's garbage collector has two components, the **reference counting collector** and the **generational garbage collector**, known as _gc module._

The reference counting algorithm is incredibly efficient and straightforward, but it cannot detect reference cycles. That is why Python has a supplemental algorithm called generational cyclic GC, that specifically deals with reference cycles.

The reference counting module is fundamental to Python and can't be disabled, whereas the cyclic GC is optional and can be invoked manually.

**Now answering your questions:**

_- Do you have an idea why Python does not release the memory automatically in this case, and therefore we have to hardcode the call to gc.collect()?_ 

> As mentioned above, because of the reference cycles, that can only be handled with generational cyclic.

_- Garbage collection has performance implications, and is supposed to be maintained by Python itself. So if we have to add this to the library code, does it mean that we should also recommend Vaex users to call gc.collect() after every (or some) aggregations?_

> Yes, defnitely! It is very often in data science to use this method, to allow freeing any memory that is no longer needed. Pandas is a lot more memory hungry, but vaex users can also benefit from this simple technique, for sure.

- What is the root cause of this problem? 

> I only went as far as Python code was related, because it seems to be where the str.byte_length() method is in a C++ file: "packages/vaex-core/src/strings.cpp"

- Can we make sure that this aggregation allocates smaller chunks of memory, that would be garbage-collected automatically?

> Yes we can! We just need to use the **gc.set_threshold()** method. It just needs to be set straight from the first import in the vaex module. However, instead of making a radical change over the vaex repo, I just did my debug with **gc.get_count()** in some check points, to allow me find the critical ones. I just focused on the issue at hand.